cpp/src/arrow/ipc/read_write_test.cc

```diff
@@ -427,6 +469,14 @@ class TestIpcRoundTrip : public ::testing::TestWithParam<MakeRecordBatch*>,
   void TearDown() { IpcTestFixture::TearDown(); }
 };
 
+class TestIpcDifferentEndian
+    : public ::testing::TestWithParam<std::tuple<MakeRecordBatch*, MakeRecordBatch*>>,
+      public IpcTestFixture {
+ public:
+  void SetUp() { IpcTestFixture::SetUp(); }
+  void TearDown() { IpcTestFixture::TearDown(); }
+};
```

Comment:
I'm not sure how difficult to implement this idea...
How about enhancing existing `IpcTestFixture` instead of adding a new fixture?

```diff
diff --git a/cpp/src/arrow/ipc/read_write_test.cc b/cpp/src/arrow/ipc/read_write_test.cc
index 923670834..1e03472b8 100644
--- a/cpp/src/arrow/ipc/read_write_test.cc
+++ b/cpp/src/arrow/ipc/read_write_test.cc
@@ -397,6 +397,9 @@ class IpcTestFixture : public io::MemoryMapFixture, public ExtensionTypesMixin {
 
     ASSERT_OK_AND_ASSIGN(result, DoLargeRoundTrip(batch, /*zero_data=*/true));
     CheckReadResult(*result, batch);
+
+    ASSERT_OK_AND_ASSIGN(result, DoEndiannessRoundTrip(batch));
+    CheckReadResult(*result, batch);
   }
   void CheckRoundtrip(const std::shared_ptr<Array>& array,
                       IpcWriteOptions options = IpcWriteOptions::Defaults(),
```

`DoEndiannessRoundTrip()` will:

* create a nonnative endianness record batch by swapping the given record batch's endianness
  * We need to implement this endianness conversion feature but how difficult...?
* run `DoStandardRoundTrip()` against the nonnative endianness record batch

BTW, should we always convert endianness in IPC? If users just relay data (A (little, generate data) -> B (big, not touch data) -> C (little, process data)), it'll be inefficient.
Or should we convert endianness explicitly out of IPC by providing `Result<std::shared_ptr<RecordBatch>> RecordBatch::EnsureEndianness(Endianness)` or something?

Reply:
Good point. In my current thought, another version of [ArrayLoader](https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/reader.cc#L107) (`LoadType` method), which is called from `LoadRecordBatchSubset`, will be implemented. One existing version is the non-endian conversion. The other is the endian conversion version. At this level, we know the type of each element. Thus, I imagine it is easy to swap endianness. What do you think?

To add a new API `Result<std::shared_ptr<RecordBatch>> RecordBatch::EnsureEndianness(Endianness)` is good idea to avoid unnecessary conversion in the above scenario. Technically, this could work correctly.
I have two questions.
1. I imagine that this API is explicitly called by a user. If so, is this change acceptable by a user? Of course, most of the scenarios (communicate between the same endianness) work well without calling this new API.
2. This new API needs to traverse all of the elements again if we need the data conversion. It may lead to additional overhead.

What do you think?  cc @pitrou